### PR TITLE
support os_profile_secrets

### PR DESCRIPTION
**Pull request by AliAllomani**
adding support for os_profile_secrets

example variable usage 

```hcl
os_profile_secrets = [{
  source_vault_id    = "/subscriptions/XXX/resourcegroups/XXX/providers/Microsoft.KeyVault/vaults/XXX"
  certificate_url    = "https://XXX.vault.azure.net/secrets/XXX/XXX"
  certificate_store  = "My"
}]
```

### CODE REVIEW

**Comment by yupwei68** [comment]
Hi @AliAllomani Thanks for your contribution. Would you mind add a test in test/fixture and add the corresponding fields in README?

**Comment by yupwei68** [comment]
Hi @AliAllomani After some changes, the test has failed too.
TestTerraformSshExample 2020-11-20T03:12:38Z command.go:100: Error: Code="KeyVaultAccessForbidden" Message="Key Vault https://test87898d10kv.vault.azure.net/secrets/test87898d10kv-cert/b97c3f7149b94b2ca952ea582d737649 either has not been enabled for deployment or the vault id provided, /subscriptions/.../resourceGroups/host87898d10-rg/providers/Microsoft.KeyVault/vaults/test87898d10kv, does not match the Key Vault's true resource id."

File: test/fixture/keyvault.tf [hunk 1]
```diff
@@ -0,0 +1,131 @@
+data "azurerm_client_config" "current" {}
+
+resource "azurerm_key_vault" "example" {
+  name                        = "testvault"
+  location                    = azurerm_resource_group.example.location
+  resource_group_name         = azurerm_resource_group.example.name
```
Comment (yupwei68) [inline test/fixture/keyvault.tf:6]: Here we use "test" for the name of the resources

File: test/fixture/keyvault.tf [hunk 2]
```diff
@@ -0,0 +1,131 @@
+data "azurerm_client_config" "current" {}
+
+resource "azurerm_key_vault" "example" {
+  name                        = "testvault"
```
Comment (yupwei68) [inline test/fixture/keyvault.tf:4]: Please add  random string in name of resources, like "test${random_id.ip_dns.hex}kv"

File: test/fixture/keyvault.tf [hunk 3]
```diff
@@ -0,0 +1,131 @@
+data "azurerm_client_config" "current" {}
+
+resource "azurerm_key_vault" "example" {
```
Comment (yupwei68) [inline test/fixture/keyvault.tf:3]: Here we use "test" instead of "example"

File: test/fixture/keyvault.tf [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+      "import",
+      "list",
+      "listissuers",
+      "managecontacts",
+      "manageissuers",
+      "setissuers",
+      "update",
+    ]
+
+    key_permissions = [
```
Comment (yupwei68) [inline test/fixture/keyvault.tf:32]: Please create a separate resource for "key_vault_access_poliy". Here only access policy for the current sp doesn't work either. We also need access policy for the vm.

File: test/fixture/keyvault.tf [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+    name  = "example"
+    phone = "0123456789"
+  }
+
+  tags = {
+    environment = "Testing"
+  }
+}
+
+resource "azurerm_key_vault_certificate" "example" {
```
Comment (yupwei68) [inline test/fixture/keyvault.tf:79]: "test" instead of "example"

File: test/fixture/main.tf [hunk 6]
```diff
@@ -99,6 +99,11 @@ module "windowsservers" {
   vnet_subnet_id      = azurerm_subnet.subnet3.id
   license_type        = var.license_type
   identity_type       = var.identity_type
+  os_profile_secrets  = [{
+                        source_vault_id     = azurerm_key_vault.example.id
+                        certificate_url     = azurerm_key_vault_certificate.example.secret_id
+                        certificate_store   = "My"
+                        }]
```
Comment (yupwei68) [inline test/fixture/main.tf:106]: Here we need to run "terraform fmt"

**Review by yupwei68 — COMMENTED** [review]
Thanks for your contribution. Please ensure the test has passed. Here's some comments.

File: test/fixture/keyvault.tf (hunk 4 above)
Comment (AliAllomani) [reply test/fixture/keyvault.tf]: key_vault_access_policy for the sp and for an assigned user identity has been added

**Comment by AliAllomani** [comment]
@yupwei68 Thanks for your comments, i've fixed the required.

**Comment by yupwei68** [comment]
Hi @AliAllomani Thanks for contribution. This PR looks almost good to me. The test has passed. Would you add this feature into the example in README to show how to use this feature?

**Comment by AliAllomani** [comment]
@yupwei68 I added the usage instructions to README, i skipped adding a full example as it would be so long to include the whole key vault resources definitions, let me know if that looks good to you or if you have any other suggestions.

File: README.md [hunk 7]
```diff
@@ -234,6 +234,36 @@ output "windows_vm_private_ips" {
...
+```
+
+Or via referencing them from `azurerm_key_vault` and `azurerm_key_vault_certificate` resources, eg.
+
+```hcl
+  os_profile_secrets = [{
+    source_vault_id   = azurerm_key_vault.test.id
+    certificate_url   = azurerm_key_vault_certificate.test.secret_id
+    certificate_store = "My"
+  }]
```
Comment (yupwei68) [inline README.md:264]: Here, could we add the features in the `Advanced Usage` example? Could we add two data source `azurerm_key_vault` and `azurerm_key_vault_certificate` and refer them here, like (here we use name `example`):
``` 
data "azurerm_key_vault" "example" {
  name                = "examplekeyvault"
  resource_group_name = azurerm_resource_group.example.name
}

data "azurerm_key_vault_certificate" "example" {
  name         = "example-kv-cert"
  key_vault_id = data.azurerm_key_vault.example.id
}
...

os_profile_secrets = [{
    source_vault_id   = data.azurerm_key_vault.example.id
    certificate_url   = data.azurerm_key_vault_certificate.example.secret_id
    certificate_store = "My"
  }]
....
```

File: README.md [hunk 8]
```diff
@@ -234,6 +234,36 @@ output "windows_vm_private_ips" {
...
+
+You can assign the values of the keys directly, eg.
+
+```hcl
+os_profile_secrets = [{
+  source_vault_id    = "/subscriptions/XXX/resourcegroups/XXX/providers/Microsoft.KeyVault/vaults/XXX"
+  certificate_url    = "https://XXX.vault.azure.net/secrets/XXX/XXX"
+  certificate_store  = "My"
+}]
+```
```
Comment (yupwei68) [inline README.md:255]: This could be omitted.

File: README.md [hunk 9]
```diff
@@ -234,6 +234,36 @@ output "windows_vm_private_ips" {
...
 
+### Virtual Machine OS Certificates / Secrets
+
+You can install custom certificates / secrets on the virtual machine from Key Vault by using the variable `os_profile_secrets`.
+
+The variable accepts a list of maps with the following keys:
+
+* source_vault_id : The ID of the Key Vault Secret which contains the encrypted Certificate.
+* certificate_url : The certificate  URL in Key Vault
+* certificate_store : The certificate store on the Virtual Machine where the certificate should be added to (Windows Only).
```
Comment (yupwei68) [inline README.md:245]: Could we add this in the 4th note of `Advanced Usage`?

File: test/fixture/main.tf [hunk 10]
```diff
@@ -64,7 +64,7 @@ module "ubuntuservers" {
   enable_ssh_key                = false
   identity_type                 = "UserAssigned"
   identity_ids                  = [azurerm_user_assigned_identity.test.id]
-  
+
```
Comment (yupwei68) [inline test/fixture/main.tf:67]: Could we remove the unnecessary changes?

Comment (yupwei68) [reply test/fixture/main.tf:67]: Here we need to add this feature into linux vm to test it as well. Please add it here:
```
os_profile_secrets = [{
    source_vault_id   = azurerm_key_vault.test.id
    certificate_url   = azurerm_key_vault_certificate.test.secret_id
    certificate_store = "My"
  }]
```

File: variables.tf [hunk 11]
```diff
@@ -220,3 +220,9 @@ variable "identity_ids" {
   type        = list(string)
   default     = []
 }
+
+variable "os_profile_secrets" {
+  description = "Specifies a list of certificates to be installed on the VM, each list item is a map with the keys source_vault_id, certificate_url, certificate_store (Windows Only)"
```
Comment (yupwei68) [inline variables.tf:225]: Here is not only for Windows, it's for linux too.
description = "Specifies a list of certificates to be installed on the VM, each list item is a map with the keys `source_vault_id`, `certificate_url`, `certificate_store` "

**Review by yupwei68 — COMMENTED** [review]
Thanks for your update! 🚀 I've added some comments in the PR. This feature is for linux as well, I've tested the feature in linux with success. Besides this, I think this PR will be good for me.

File: variables.tf (hunk 11 above)
Comment (AliAllomani) [reply variables.tf:225]: "Windows Only" here is referring to the key "certificate_store" only, but i can see that this may create some confusion that way that the whole "os_profile_secrets" is windows only, i will remove it from here and keep it in readme only.

File: test/fixture/main.tf (hunk 10 above)
Comment (AliAllomani) [reply test/fixture/main.tf]: updated

File: README.md (hunk 9 above)
Comment (AliAllomani) [reply README.md]: updated

File: README.md (hunk 7 above)
Comment (AliAllomani) [reply README.md]: updated

**Review by yupwei68 — APPROVED** [review]
LGTM!
